### PR TITLE
Expose new core functionality to Python

### PR DESCRIPTION
**Pull request by soltanmm**
cc #1616, #2509, #2578, #2636

### CODE REVIEW

File: src/python/grpcio/grpc/_adapter/_types.py [hunk 1]
```diff
@@ -321,6 +354,28 @@ def create_call(self, completion_queue, method, host, deadline=float('+inf')):
     """
     return None
 
+  @abc.abstractmethod
+  def connectivity_state(self, try_to_connect=False):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_adapter/_types.py]: Should there not be separate check_connectivity_state and watch_connectivity_state methods on this object?

Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_adapter/_types.py]: Please avoid optional parameters on internal APIs. It's no trouble to us to always be explicit about passing the parameter all the time.

File: src/python/grpcio/grpc/_adapter/_types.py [hunk 2]
```diff
@@ -138,39 +161,40 @@ class OpArgs(collections.namedtuple(
...
 
   @staticmethod
   def send_initial_metadata(initial_metadata):
-    return OpArgs(OpType.SEND_INITIAL_METADATA, initial_metadata, None, None, None)
+    return OpArgs(OpType.SEND_INITIAL_METADATA, initial_metadata, None, None, None, 0)
 
   @staticmethod
-  def send_message(message):
-    return OpArgs(OpType.SEND_MESSAGE, None, None, message, None)
+  def send_message(message, flags=0):
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_adapter/_types.py]: I'm wary of optional parameters on internal APIs. Can this be required?

File: src/python/grpcio/grpc/_adapter/_types.py (hunk 1 above)
Comment (soltanmm) [inline src/python/grpcio/grpc/_adapter/_types.py]: W.r.t. separate methods: I was mirroring the C API. Would you prefer a departure?

Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_adapter/_types.py]: When I look at the C API I see two separate methods (grpc_channel_watch_connectivity_state and grpc_channel_check_connectivity_state) - am I misreading the API?

Comment (soltanmm) [inline src/python/grpcio/grpc/_adapter/_types.py]: Ah, I misunderstood. With respect to `grpc_channel_watch_connectivity_state`, I thought we discussed in a past meeting that we didn't want to use the queue?

Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_adapter/_types.py]: We want to use it; we're just not going to be directly exposing to applications.

File: src/python/grpcio/grpc/_adapter/_c/types/channel.c [hunk 3]
```diff
@@ -36,10 +36,14 @@
 #define PY_SSIZE_T_CLEAN
 #include <Python.h>
 #include <grpc/grpc.h>
+#include <grpc/support/alloc.h>
 
 
 PyMethodDef pygrpc_Channel_methods[] = {
     {"create_call", (PyCFunction)pygrpc_Channel_create_call, METH_KEYWORDS, ""},
+    {"connectivity_state", (PyCFunction)pygrpc_Channel_check_connectivity_state, METH_KEYWORDS, ""},
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_adapter/_c/types/channel.c]: Can we name this "check_connectivity_state"?

**Comment by nathanielmanistaatgoogle** [comment]
Looks good modulo the one nitpick and pending test results.

File: src/python/grpcio/grpc/_adapter/_c/types/channel.c (hunk 3 above)
Comment (soltanmm) [inline src/python/grpcio/grpc/_adapter/_c/types/channel.c]: Done - that was an oversight.

**Pull request merged**